include/envoy/http/header_map.h

```diff
@@ -601,6 +627,16 @@ class HeaderMap {
    */
   virtual void dumpState(std::ostream& os, int indent_level = 0) const PURE;
 
+  /**
+   * @return true if header field-name case-preservation is enabled in map, false otherwise.
+   */
+  virtual bool isHeadersCasePreservationEnabled() const PURE;
+
+  /**
+   * Enables / Disables the header field-name case-preservation in the map.
+   */
+  virtual void enableHeadersCasePreservation(bool enabled) PURE;
```

Comment:
I prefer `setHeaderKeysCasePreserved(bool)` but doesn't matter that much (I am still not convinced this needs to be on the map, only the entry)